google/cloud/pubsub/internal/batching_publisher_connection_test.cc

```diff
@@ -115,7 +113,11 @@ TEST(BatchingPublisherConnectionTest, DefaultMakesProgress) {
   publisher->Flush({});
   // Use the CQ threads to satisfy the AsyncPull future, like we do in the
   // normal code.
-  background.cq().RunAsync([&async] { async.PopFront().set_value(); });
+  background.cq().RunAsync([&async, publish_count] {
+    for (auto i = 0; i != publish_count; ++i) {
```

Comment:
This happens in the same thread as the `AsyncPublish()`, because `background` creates a single thread, otherwise you would need some synchronization.

Reply:
If we are assuming there are multiple threads involved, I think atomics are not guaranteed to save the day.  The change may not be visible when this thread runs, so you may get "0" when loading `publish_count`, even if it was incremented twice and "async" contains two elements:

https://stackoverflow.com/questions/42003798/how-do-i-make-memory-stores-in-one-thread-promptly-visible-in-other-threads

And therefore "Bad Things[tm]" happen (probably a deadlock).  

Well, it is complicated, again.  It probably works, but for other obscure reasons: `RunAsync()` only runs once `Flush()` completes and that only after the two `PublishAsync()` complete.  `RunAsync()` uses mutexes to push the the lambda to the CQ thread.  Those mutexes will preserve the "happens before" relationships and make the changes to `publish_count` (and anything else) visible to that other thread...  But the atomic itself was not enough.

If we are going to assume synchronization is needed, then maybe just add a new `.PopFrontSafe() -> optional<promise<T>>` or `.Empty()` to the `AsyncSequencer` and use it instead of a separate counter.

Thoughts?